File: Chapter 03/Python/section_3_3_2_ocp/after/midfielder.py
from .play import Play

"""
OCP SOLUTION: Closed for Modification.
* ARCHITECTURE NOTE: Look at the simplicity here! The massive if/elif chain 
is completely gone. The Midfielder no longer decides HOW a play works; 
it simply tells the play to execute itself. 
* This class is now "Closed for Modification"—we can add 100 new plays to 
the playbook and we will never have to touch this file again.
"""
class Midfielder:
    def execute_play(self, play: 'Play'):
        """
        Executes any play that follows the Play contract.
        """
        # The Midfielder delegates the work to the Strategy object.
        play.execute()

File: Chapter 03/JS/section_3_3_2_ocp/after/midfielder.js
/**
 * OCP SOLUTION: Closed for Modification.
 * * ARCHITECTURE NOTE: This class is now perfectly "Closed for modification." 
 * Notice that the previous if/else chain is completely GONE. 
 * * The Midfielder no longer cares WHAT the specific play is or how it functions. 
 * It only cares that the play object adheres to the "contract" by providing 
 * an execute() method. We can now add infinite new plays to the system 
 * without ever opening or editing this file again.
 */
class Midfielder {
    /**
     * Executes any play dynamically.
     * @param {Object} play - An object that must implement an execute() method.
     */
    executePlay(play) {
        // ARCHITECTURE NOTE: In JavaScript, we use 'Duck Typing' or explicit 
        // checks to ensure the object follows our expected contract.
        if (typeof play.execute !== 'function') {
            throw new Error("Play must implement an execute() method!");
        }
        
        // The Midfielder delegates the work to the Strategy object.
        play.execute();
    }
}

module.exports = Midfielder;